File: .claude/hooks/block_illegal_path_edits.ts
#!/usr/bin/env bun
import { readFileSync } from "node:fs";
import path from "node:path";

function emitDeny(reason: string): void {
  console.log(
    JSON.stringify({
      hookSpecificOutput: {
        hookEventName: "PreToolUse",
        permissionDecision: "deny",
        permissionDecisionReason: reason,
      },
    }),
  );
}

async function main(): Promise<void> {
  const raw = readFileSync(0, "utf8").trim();
  if (!raw) return;

  let payload: Record<string, unknown>;
  try {
    payload = JSON.parse(raw) as Record<string, unknown>;
  } catch {
    return;
  }

  const toolInput = (payload.tool_input ?? {}) as Record<string, unknown>;
  const filePath = String(toolInput.file_path ?? "").trim();
  if (!filePath) return;

  const projectRoot = path.resolve(process.env.CLAUDE_PROJECT_DIR ?? process.cwd());
  const target = path.resolve(projectRoot, filePath);
  const rel = path.relative(projectRoot, target);
  const relPosix = rel.split(path.sep).join("/");

  if (rel.startsWith("..") || path.isAbsolute(rel)) {
    emitDeny(`Blocked edit outside repository root: ${filePath}`);
    return;
  }

  if (relPosix === ".git" || relPosix.startsWith(".git/")) {
    emitDeny("Blocked edits under .git/");
    return;
  }

  if (relPosix === ".claude/settings.local.json") {
    emitDeny("Blocked edits to local-only Claude settings (.claude/settings.local.json)");
    return;
  }

  if (relPosix.startsWith(".claude/worktrees/")) {
    emitDeny("Blocked edits inside .claude/worktrees/");
    return;
  }

  if (relPosix.startsWith("openspec/schemas/")) {
    emitDeny("Blocked edits to openspec/schemas/ (use openspec schema commands instead)");
  }
}

main().catch(() => {
  // no-op on hook failures
});

File: .claude/hooks/block_illegal_path_edits.py
#!/usr/bin/env python3
"""PreToolUse hook: block illegal path edits for ANVIL constraints."""

from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path


DERIVED_STATE_RE = re.compile(r"^work/features/[^/]+/state\.yaml$")


def emit_deny(reason: str) -> None:
    payload = {
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": "deny",
            "permissionDecisionReason": reason,
        }
    }
    print(json.dumps(payload))


def normalize_path(project_root: Path, raw_path: str) -> Path:
    candidate = Path(raw_path)
    if not candidate.is_absolute():
        candidate = project_root / candidate
    return candidate.resolve(strict=False)


def path_outside_project(project_root: Path, target: Path) -> bool:
    try:
        target.relative_to(project_root)
        return False
    except ValueError:
        return True


def main() -> int:
    raw = sys.stdin.read().strip()
    if not raw:
        return 0

    try:
        payload = json.loads(raw)
    except json.JSONDecodeError:
        return 0

    tool_input = payload.get("tool_input") or {}
    file_path = str(tool_input.get("file_path") or "").strip()
    if not file_path:
        return 0

    project_root = Path(os.environ.get("CLAUDE_PROJECT_DIR") or os.getcwd()).resolve(strict=False)
    target = normalize_path(project_root, file_path)

    if path_outside_project(project_root, target):
        emit_deny(f"Blocked edit outside repository root: {file_path}")
        return 0

    rel = target.relative_to(project_root).as_posix()

    if rel == ".git" or rel.startswith(".git/"):
        emit_deny("Blocked edits under .git/")
        return 0

    if rel == ".claude/settings.local.json":
        emit_deny("Blocked edits to local-only Claude settings (.claude/settings.local.json)")
        return 0

    if rel.startswith(".claude/worktrees/"):
        emit_deny("Blocked edits inside .claude/worktrees/")
        return 0

    if DERIVED_STATE_RE.match(rel):
        emit_deny("Blocked manual edits to derived state cache: work/features/<id>/state.yaml")
        return 0

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
